File: app/graphql/object_types/channel.rb
# frozen_string_literal: true

ObjectTypes::Channel = GraphQL::ObjectType.define do
  name "Channel"

  implements GraphQL::Relay::Node.interface

  global_id_field :id

  field :annictId, !types.Int do
    resolve ->(obj, _args, _ctx) {
      obj.id
    }
  end

  connection :programs, ObjectTypes::Program.connection_type do
    resolve ->(obj, _args, _ctx) {
      ForeignKeyLoader.for(Program, :channel_id).load([obj.id])
    }
  end

  field :channelGroup, !ObjectTypes::ChannelGroup do
    resolve ->(obj, _args, _ctx) {
      RecordLoader.for(ChannelGroup).load(obj.channel_group_id)
    }
  end

  field :scChid, !types.Int do
    resolve ->(obj, _args, _ctx) {
      obj.sc_chid
    }
  end

  field :name, !types.String
  field :published, !types.Boolean
end

File: app/frontend/javascripts/v3/models/Channel.ts
import { ApplicationModel } from './ApplicationModel'
import { Program, Work } from '../models'

export class Channel extends ApplicationModel {
  public annictId: number
  public name: string
  public programs: Program[]

  public constructor(node) {
    super()
    this.annictId = node.annictId
    this.name = node.name
  }

  public setProgramsOfWork(work: Work) {
    this.programs = work.programs.filter(program => {
      return program.channel.annictId === this.annictId
    })
  }
}
